source/readonly-deep.d.ts

```diff
@@ -0,0 +1,57 @@
+import {Primitive} from './basic';
+
+/**
+Convert `object`s, `Map`s, `Set`s, and `Array`s into immutable stuctures recursively. Converts all properties/contents into immutable structures recusively.
+
+This is useful when a deeply nested structure needs to be exposed as completely immutable, like for example an imported JSON module.
+
+@example
+```
+// data.json
+{
+	"foo": ["bar"]
+}
+
+// main.ts
+import {ReadonlyDeep} from 'type-fest';
+import dataJson = require('./data.json');
+
+const data: ReadonlyDeep<typeof dataJson> = dataJson;
+
+export default data;
+
+// test.ts
+import data from './main';
+
+data.foo.push('bar');
+//=> error TS2339: Property 'push' does not exist on type 'readonly string[]'
+```
+*/
+export type ReadonlyDeep<T> = T extends Primitive | ((...arguments: any[]) => unknown)
```

Comment:
Should we maybe move `((...arguments: any[]) => unknown)` be a `AnyFunction` type?

Reply:
Not sure. If we do this, why not simply using `Function`?

Reply:
Good point. Never mind.